lighthouse-core/gather/computed/dependency-graph/estimator/estimator.js

```diff
@@ -0,0 +1,291 @@
+  });
+
+  return grouped;
+}
+
+class Estimator {
+  /**
+   * @param {!Node} graph
+   * @param {{rtt: number, throughput: number, fallbackTTFB: number,
+   *    maximumConcurrentRequests: number}=} options
+   */
+  constructor(graph, options) {
+    this._graph = graph;
+    this._options = Object.assign(
+      {
+        rtt: DEFAULT_RTT,
+        throughput: DEFAULT_THROUGHPUT,
+        fallbackTTFB: DEFAULT_FALLBACK_TTFB,
+        maximumConcurrentRequests: DEFAULT_MAXIMUM_CONCURRENT_REQUESTS,
+      },
+      options
+    );
+
+    this._rtt = this._options.rtt;
+    this._throughput = this._options.throughput;
+    this._fallbackTTFB = this._options.fallbackTTFB;
+    this._maximumConcurrentRequests = Math.min(
+      TcpConnection.maximumSaturatedConnections(this._rtt, this._throughput),
+      this._options.maximumConcurrentRequests
+    );
+  }
+
+  /**
+   * @param {!WebInspector.NetworkRequest} record
+   * @return {number}
+   */
+  static getTTFB(record) {
+    const timing = record._timing;
+    return (timing && timing.receiveHeadersEnd - timing.sendEnd) || Infinity;
+  }
+
+  _initializeNetworkRecords() {
+    this._networkRecords = [];
+
+    this._graph.getRootNode().traverse(node => {
+      if (node.type === Node.TYPES.NETWORK) {
+        this._networkRecords.push(node.record);
+      }
+    });
+  }
+
+  _initializeNetworkConnections() {
+    const connections = new Map();
+    const recordsByConnection = groupBy(
+      this._networkRecords,
+      record => record.connectionId
+    );
+
+    for (const [connectionId, records] of recordsByConnection.entries()) {
+      const isTLS = TLS_SCHEMES.includes(records[0].parsedURL.scheme);
+
+      // We'll approximate how much time the server for a connection took to respond after receiving
+      // the request by computing the minimum TTFB time for requests on that connection.
+      //    TTFB = one way latency + server response time + one way latency
+      // Even though TTFB is greater than server response time, the RTT is underaccounted for by
+      // not varying per-server and so the difference roughly evens out.
+      // TODO(patrickhulce): investigate a way to identify per-server RTT
+      let estimatedResponseTime = records.reduce(
+        (min, record) => Math.min(min, Estimator.getTTFB(record)),
+        Infinity
+      );
+
+      // If we couldn't find a TTFB for the requests, use the fallback TTFB instead.
+      if (!Number.isFinite(estimatedResponseTime)) {
+        estimatedResponseTime = this._fallbackTTFB;
+      }
+
+      const connection = new TcpConnection(
+        this._rtt,
+        this._throughput,
+        estimatedResponseTime,
+        isTLS
+      );
+
+      connections.set(connectionId, connection);
+    }
+
+    this._connections = connections;
+    return connections;
+  }
+
+  _initializeAuxiliaryData() {
+    this._nodeTiming = new Map();
+    this._nodesCompleted = new Set();
+    this._nodesInProcess = new Set();
```

Comment:
in the same process? if not, may be a bit confusing naming-wise. `nodesInProgress`?

Reply:
hahaha, good point, done